lms/services/grouping.py

```diff
@@ -0,0 +1,68 @@
+from lms.models import CanvasSection, Grouping
+from lms.models._hashed_id import hashed_id
+
+
+class GroupingService:
+    def __init__(self, db, application_instance_service, course_service):
+        self._db = db
+        self._application_instance = application_instance_service.get()
+        self._course_service = course_service
+
+    def upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
+                type=grouping.type,
+            )
+            .one_or_none()
+        )
+        if not db_grouping:
+            self._db.add(grouping)
+        else:
+            # Update any fields that might have changed
+            db_grouping.lms_name = grouping.name
+            db_grouping.extra = grouping.extra
```

Comment:
I'd need a separate db_grouping or similar to do the update part, but witth the try: block it might be unbounded so I think it will complicate things.

`.lms_name = .name` was wrong in my code, should be `lms_name` in both ends